### PR TITLE
Enable counter_enabled.py to support batch mode

### PR DESCRIPTION
**Pull request by nabheet**
##### SUMMARY

Enable counter_enabled.py to support batch playbook executions using the serial tag in plays. Currently, the host counter gets reset at the beginning of every task. However, during batch executions we want it to keep track of the previous batch executions and print the host counter based on the previous runs. This proposal keeps track of how many servers have been updated in previous batches and starts the host counter at that tracked value.

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

counter_enabled

##### ADDITIONAL INFORMATION

When executing a playbook in batch mode, i.e., with the serial attribute like

```
- hosts: all
  gather_facts: no
  serial:
    - 3
    - 15%
    - 20%
    - 35%
    - 55%
    - 90%
    - 100%
  tasks:
    - name: Ping Hello!
      ping:
        data: "Hello!!!!"
```

The callback plugin `counter_enabled` does not keep track of batch size. This is a bug fix to report the host counter in the current batch correctly.

Please note that the server order in PLAY RECAP is not relevant to the issue.
Before change:
```
PLAY [all] *********************************************************************************************

TASK 1/1 [Ping Hello!] ******************************************************************************************
ok: 1/9 [server01]
ok: 2/9 [server02]
ok: 3/9 [server03]

PLAY [all] *********************************************************************************************

TASK 2/1 [Ping Hello!] ******************************************************************************************
ok: 1/9 [server04]

PLAY [all] *********************************************************************************************

TASK 3/1 [Ping Hello!] ******************************************************************************************
ok: 1/9 [server05]

PLAY [all] *********************************************************************************************

TASK 4/1 [Ping Hello!] ******************************************************************************************
ok: 1/9 [server06]
ok: 2/9 [server07]
ok: 3/9 [server08]

PLAY [all] *********************************************************************************************

TASK 5/1 [Ping Hello!] ******************************************************************************************
ok: 1/9 [server09]

PLAY RECAP ******************************************************************************************************
server01               : ok=1    changed=0    unreachable=0    failed=0    rescued=0    ignored=0
server02               : ok=1    changed=0    unreachable=0    failed=0    rescued=0    ignored=0
server03               : ok=1    changed=0    unreachable=0    failed=0    rescued=0    ignored=0
server04               : ok=1    changed=0    unreachable=0    failed=0    rescued=0    ignored=0
server05               : ok=1    changed=0    unreachable=0    failed=0    rescued=0    ignored=0
server06               : ok=1    changed=0    unreachable=0    failed=0    rescued=0    ignored=0
server07               : ok=1    changed=0    unreachable=0    failed=0    rescued=0    ignored=0
server08               : ok=1    changed=0    unreachable=0    failed=0    rescued=0    ignored=0
server09               : ok=1    changed=0    unreachable=0    failed=0    rescued=0    ignored=0
```
After change:
```
PLAY [all] *********************************************************************************************

TASK 1/1 [Ping Hello!] ******************************************************************************************
ok: 1/9 [server01]
ok: 2/9 [server02]
ok: 3/9 [server03]

PLAY [all] *********************************************************************************************

TASK 1/1 [Ping Hello!] ******************************************************************************************
ok: 4/9 [server04]

PLAY [all] *********************************************************************************************

TASK 1/1 [Ping Hello!] ******************************************************************************************
ok: 5/9 [server05]

PLAY [all] *********************************************************************************************

TASK 1/1 [Ping Hello!] ******************************************************************************************
ok: 6/9 [server06]
ok: 7/9 [server07]
ok: 8/9 [server08]

PLAY [all] *********************************************************************************************

TASK 1/1 [Ping Hello!] ******************************************************************************************
ok: 9/9 [server09]

PLAY RECAP ******************************************************************************************************
server01               : ok=1    changed=0    unreachable=0    failed=0    rescued=0    ignored=0
server02               : ok=1    changed=0    unreachable=0    failed=0    rescued=0    ignored=0
server03               : ok=1    changed=0    unreachable=0    failed=0    rescued=0    ignored=0
server04               : ok=1    changed=0    unreachable=0    failed=0    rescued=0    ignored=0
server05               : ok=1    changed=0    unreachable=0    failed=0    rescued=0    ignored=0
server06               : ok=1    changed=0    unreachable=0    failed=0    rescued=0    ignored=0
server07               : ok=1    changed=0    unreachable=0    failed=0    rescued=0    ignored=0
server08               : ok=1    changed=0    unreachable=0    failed=0    rescued=0    ignored=0
server09               : ok=1    changed=0    unreachable=0    failed=0    rescued=0    ignored=0
```

### CODE REVIEW

**Comment by nabheet** [comment]
I just noticed that the task counter doesn't get reset. 

probably need to add this to on_play_start:
```
self._task_counter = 1
```

**Comment by russoz** [comment]
Hi @nabheet ,

Thank you for the contribution! Please add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to).

I am not very familiar with callback plugins code, but AFAICT it looks good.

**Comment by nabheet** [comment]
I don't know why this is broken. Do I just need run the build again, somehow?

File: changelogs/fragments/3709-support-batch-mode.yml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - counter_enabled - fix output to correctly display host and task counters (https://github.com/ansible-collections/community.general/pull/3709).
```
Comment (felixfontein) [inline changelogs/fragments/3709-support-batch-mode.yml:2]:
```suggestion
  - counter_enabled callback plugin - fix output to correctly display host and task counters (https://github.com/ansible-collections/community.general/pull/3709).
```

**Comment by felixfontein** [comment]
> I don't know why this is broken. Do I just need run the build again, somehow?

It looks like a random network failure. I've triggered the failed run, it should pass now. You can also trigger a re-run yourself by adding `/rebuild_failed` as a new comment.

File: changelogs/fragments/3709-support-batch-mode.yml (hunk 1 above)
Comment (nabheet) [reply changelogs/fragments/3709-support-batch-mode.yml]: Thank you!

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
CI error is unrelated, so I'm merging anyway. I'll come up with a fix in another PR.

**Comment by felixfontein** [comment]
@nabheet thanks for improving this plugin!
@russoz thanks for reviewing!

**Comment by nabheet** [comment]
@russoz and @felixfontein Thank you! And this was my first PR here! Hopefully first of many more.

**Pull request merged**